bublecs/bublecs.h

```diff
@@ -3,188 +3,309 @@
-			return a.type.Equal(type);
-		});
-
-		if (archtype != archtypes.end()) {
-			return *archtype;
-		}
-
-		Archtype newArchtype;
-		newArchtype.columns.resize(type.types.size());
-		newArchtype.type = type;
-		newArchtype.rowsCount = 0;
-		archtypes.emplace_back(newArchtype);
-		return archtypes.back();
-	}
-
-	template <typename... Components>
-	void AddToArchtypes(Entity& entity, Components&&... components)
-	{
-		Archtype& archtype = FindOfCreateArchtype(entity.type);
-		entity.row = archtype.rowsCount;
-		archtype.AddRow(std::forward<std::remove_reference_t<Components>>(components)...);
-	}
-
-	template <typename Tuple, typename Func, size_t... I>
-	void for_each_impl(Func f, Tuple t, int rows_count, std::index_sequence<I...>)
-	{
-		for (int i = 0; i < rows_count; i++) {
-			f(std::get<I>(t)[i]...);
-		}
-	}
-
-	template <typename... Components, typename Func>
-	void for_each(Func f)
-	{
-		CompositeType type = GetCompositeType<Components...>();
-
-		for (Archtype& a : archtypes) {
-			if (a.type.IsSupersetOf(type)) {
-				std::vector<int> componentIndices(type.types.size());
-
-				for (int i = 0; i < componentIndices.size(); i++) {
-					componentIndices[i] = a.type.IndexOf(type.types[i].id);
-				}
-
-				auto selectedColumns = std::make_tuple(
-					(Components*)a.columns[a.type.IndexOf(GetTypeId<Components>())].data()...);
-
-				static constexpr auto size = std::tuple_size<decltype(selectedColumns)>::value;
-
-				for_each_impl(f, selectedColumns, a.rowsCount, std::make_index_sequence<size>{});
-			}
-		}
-	}
+    std::unordered_map<ECSId, Entity> entities;
+    std::unordered_map<ECSId, Archetype> archetypes;
+
+    template <typename... Components>
+    ECSId CreateEntity(Components&&... components)
+    {
+        Entity entity;
+        entity.id = GenerateNewEntityId();
+        entity.type = GetCompositeType<Components...>();
+
+        AddToArchetypes(entity, std::forward<Components>(components)...);
+
+        entities.emplace(entity.id, entity);
+
+        return entity.id;
+
+    }
+
+    std::pair<ECSId, Archetype&> FindOrCreateArchtype(const CompositeType& type)
+    {
+        auto archetype = std::find_if(archetypes.begin(), archetypes.end(), [&](auto& p) {
+            auto& [id, arch] = p;
+            return  arch.type.Equal(type);
+        });
+
+        if (archetype != archetypes.end()) {
+            auto [id, arch] = *archetype;
+            return {id, arch};
+        }
+
+        Archetype newArchtype;
+        newArchtype.columns.resize(type.Size());
+        newArchtype.type = type;
+        newArchtype.rowsCount = 0;
+        ECSId arhc_id = GenerateNewArchetypeId();
+        archetypes.emplace(arhc_id, newArchtype);
+        return {arhc_id, archetypes[arhc_id]};
+    }
+
+    template <typename... Components>
+    void AddToArchetypes(Entity& entity, Components&&... components)
+    {
+        auto [id, archtype] = FindOrCreateArchtype(entity.type);
+        entity.archetype_id = id;
```

Comment:
I think we should store archtypes in hashtable also, so we don't need archtype_id at all. We could use entity composite type as a key.

Reply:
CompositeType as a key would require a hash function ( or some cast to simple type with hash, maybe), with just id it's easier